### PR TITLE
adding the skipPhases option to the cluster definitions, structs and openapi schema

### PR DESCRIPTION
**Pull request by drew-viles**
This is probably missing a load of stuff!

I'm more wanting guidance on this as we need to have a way of implementing the skipProxy bits.

I've been through the testing steps but they look like they are from the old world style of testing all of this stuff! 
Is there any up-to-date guides I can go through to make sure this is fully tested?

Also I'm thinking that, for now, we should surface an option to the customer/user on whether they want cilium and if they do, automatically enable `skipPhase: ["kube-proxy"]`. something like "enable Cilium: recommended by unikorn cloud!".

I appreciate this may be missing a bunch of bits but it's a complex code base and it' my first run at it since the new design :-D

### CODE REVIEW

**Comment by drew-viles** [comment]
Also, I'm guessing I'll need to bump the helm chart too?

**Review by spjmurray — CHANGES_REQUESTED** [review]
There is no point in exposing this at the API level as I seriously doubt any user, either from a UI or CLI perspective will ever use this, and it'll expose a method for users to inject any old whack values and cause unmitigated mayhem.

To be honest, given what you need it to do, and because it should work, precedence and all, you literally need to set:

```yaml
spec:
  versions:
  - parameters:
    - name: controlPlane.kubeadmSkipPhases[0]
      value: foo
```

In `charts/kubernetes/templates/applications.yaml` for the `cluster-openstack` application.

Pow, 2 line change 💥

**Comment by drew-viles** [comment]
Yeah I'm more than happy with that.

I wasn't sure if we would mandate it as part of unikorn-cloud though hence why I thought I'd surface it as an option. HI'm more than happy to go down the route of, "If you're using unikorn cloud then you're using Cilium, and since you're using that, you're not having Kube-Proxy"

File: charts/kubernetes/templates/applications.yaml [hunk 1]
```diff
@@ -162,6 +162,9 @@ spec:
   tags:
   - infrastructure
   versions:
+  - parameters:
```
Comment (spjmurray) [inline charts/kubernetes/templates/applications.yaml:165]: Close, `parameters` is a key under a version, see the use of it else where, and try keep the ordering of the properties vaguely consistent, my OCD eyes will pick out odd formatting!

Comment (drew-viles) [reply charts/kubernetes/templates/applications.yaml:165]: haha no worries. to be honest I was pushing to do some testing but then colima cocked up due to a bug so might be worth ignoring this PR for now until testing is done ;-)

Comment (drew-viles) [reply charts/kubernetes/templates/applications.yaml:165]: Also face palming at my error there. It's quite, quite clear it's part of the version itsel and not another list within the object 🤦

**Comment by spjmurray** [comment]
Okay, so this appears to actually apply correctly:

Screenshot from 2024-07-30 10-40-34

But it's not actually provisioning anything sadly, just investigating now... could just be pebcak!

**Comment by spjmurray** [comment]
There we go...
```
[   12.897914] cloud-init[988]: [2024-07-30 09:39:44] [init] Using Kubernetes version: v1.30.3
[   12.901107] cloud-init[988]: [2024-07-30 09:39:44] invalid phase name: addons/kube-proxy
```

**Comment by drew-viles** [comment]
> There we go...
> 
> ```
> [   12.897914] cloud-init[988]: [2024-07-30 09:39:44] [init] Using Kubernetes version: v1.30.3
> [   12.901107] cloud-init[988]: [2024-07-30 09:39:44] invalid phase name: addons/kube-proxy
> ```

que? Hold please. have I fat fingered something there then...

**Comment by drew-viles** [comment]
🤦 It's addon, not addons 😮‍💨  One tick

**Comment by drew-viles** [comment]
and THAT was the evidence of someone messing it up!

**Comment by spjmurray** [comment]
I _will_ always catch you out for not testing!

**Comment by drew-viles** [comment]
I'm trying to test now - You're just too quick for me :-D

**Comment by spjmurray** [comment]
I'm already deploying... catch up

**Comment by drew-viles** [comment]
Oh I know you are, but it's prudent for me to be able to test so you don't have to every time ;-)

**Comment by spjmurray** [comment]
Okay so, Cilium seems broken now:

Screenshot from 2024-07-30 11-11-04

and 

Screenshot from 2024-07-30 11-11-18

Which I will leave up to you, once you get that far...

**Comment by drew-viles** [comment]
This didn't happen when testing the charts directly. Yeah leave it with me

**Comment by drew-viles** [comment]
ok, this is to do with [this issue](https://github.com/kubernetes/cloud-provider-openstack/issues/2611)

Which *should* be fixed here: https://github.com/unikorn-cloud/kubernetes/blob/main/pkg/provisioners/helmapplications/openstackcloudprovider/provisioner.go#L153

**Comment by spjmurray** [comment]
Odd how Nick found this ages ago and it's still a problem :D

**Comment by spjmurray** [comment]
For context, because it was discussed via other forums...

The problem with this not working is actually that cilium needs to be [told the IP address of the API server](https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#kubeproxy-free).  Obviously we shouldn't even need to know this level of detail to install something so there is no support for that.  Exposing this detail will happen in a separate piece of work.   CAPI needs to know this API information in a generic manner in order to create the Kubernetes config, but that would require smarts.  This would cause additional liable-to-break cohesion between us and 3rd party components, especially when you consider the joys of CRDs and the hacks required to actually remove a workload pool due to both CAPI and Argo being broken in different ways.

What we do have is access to the Kubernetes config secret already, which is nice and generic, and we even have to grab it to talk to the cluster in the first place, so perhaps adding some remote cluster foo to the context should suffice...

**Comment by spjmurray** [comment]
```diff
diff --git a/pkg/provisioners/helmapplications/cilium/provisioner.go b/pkg/provisioners/helmapplications/cilium/provisioner.go
index c4a005b..706e12f 100644
--- a/pkg/provisioners/helmapplications/cilium/provisioner.go
+++ b/pkg/provisioners/helmapplications/cilium/provisioner.go
@@ -19,7 +19,10 @@ package cilium
 
 import (
        "context"
+       "fmt"
 
+       coreclient "github.com/unikorn-cloud/core/pkg/client"
+       coreerrors "github.com/unikorn-cloud/core/pkg/errors"
        "github.com/unikorn-cloud/core/pkg/provisioners/application"
        "github.com/unikorn-cloud/core/pkg/provisioners/util"
        unikornv1 "github.com/unikorn-cloud/kubernetes/pkg/apis/unikorn/v1alpha1"
@@ -52,8 +55,26 @@ func (p *Provisioner) Values(ctx context.Context, _ *string) (interface{}, error
                operatorValues["replicas"] = cluster.Spec.ControlPlane.Replicas
        }
 
+       // We run in sans-kube-proxy mode, as it's faster and doesn't involve
+       // iptables ;-), for that we need to specify the Kubernetes API endpoint
+       // as per https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free.
+       // This information is propagated as part of the cluster metadata in the
+       // context.  Raise an error if not set, as this should only be used in
+       // the context of a remote cluster.
+       clusterContext, err := coreclient.ClusterFromContext(ctx)
+       if err != nil {
+               return nil, err
+       }
+
+       if clusterContext.Host == "" || clusterContext.Port == "" {
+               return nil, fmt.Errorf("%w: missing cluster host:port", coreerrors.ErrInvalidContext)
+       }
+
        values := map[string]interface{}{
-               "operator": operatorValues,
+               "operator":             operatorValues,
+               "kubeProxyReplacement": "true",
+               "k8sServiceHost":       clusterContext.Host,
+               "k8sServicePort":       clusterContext.Port,
                "hubble": map[string]interface{}{
                        "relay": map[string]interface{}{
                                "nodeSelector": util.ControlPlaneNodeSelector(),
```

**Comment by drew-viles** [comment]
We have a winner!

image

**Review by spjmurray — APPROVED** [review]

**Pull request merged**